### PR TITLE
Add intrinsics for signed and unsigned multiply-high

### PR DESCRIPTION
**Pull request by gretay-js**
Add intrinsics to compute the high 64 bit of the the product of two 64-bit integers.
Signed version of this operation is already present in Cmm as `Cmulh` (and used for optimized division), but not exposed as a builtin. This PR adds intrinsics for both signed and unsigned versions. To handle them uniformly, it extend `Cmulh` with a boolean argument `signed` and implements the unsigned operation in all targets.

Notes:
1) There is no need for an unsigned version of low 64-bits multiplication (that is the "normal" truncated multiplication), because the low part of the product is the same for signed and unsigned operands.
2)  Multiply-high for 32 bit integers can be implemented using int64, so I didn't add intrinsics for it, but it wouldn't be hard, as the emitters of 32-bit architectures correctly handle both versions of `Mach.Imulh` operation, even though that code is not in use, because int64 intrinsics on 32-bit targets use Cextcall.

### CODE REVIEW

File: backend/arm/emit.mlp
Comment (xclerc) [inline backend/arm/emit.mlp:198]: Shouldn't we have a line matching `Imulh { signed = false }`?

File: backend/arm/emit.mlp
Comment (xclerc) [inline backend/arm/emit.mlp:722]: Should this match also have the `!arch < ARMv6` guard?

File: backend/arm/proc.ml
Comment (xclerc) [inline backend/arm/proc.ml:300]: Shouldn't this match also have the `!arch < ARMv6` guard?

File: backend/arm/proc.ml
Comment (xclerc) [inline backend/arm/proc.ml:323]: Shouldn't this match also have the `!arch < ARMv6` guard?

File: backend/arm/proc.ml
Comment (xclerc) [inline backend/arm/proc.ml:333]: Shouldn't this match also have the `!arch < ARMv6` guard?

File: backend/mach.ml
Comment (xclerc) [inline backend/mach.ml:261]: (Use `left`/`right`, like the surrounding code.)

File: backend/printcmm.ml
Comment (xclerc) [inline backend/printcmm.ml:167]: In `Cfg`, `u` is added for unsigned while `s` is added here fore signed;
shouldn't we just always have a suffix to enforce uniformity?

Comment (gretay-js) [reply backend/printcmm.ml:167]: Ah, thanks, this one was a mistake. I was trying to match upstream output for the signed version.

File: backend/mach.ml
Comment (gretay-js) [reply backend/mach.ml:261]: Fixed.

**Review by xclerc — COMMENTED** [review]
I am fairly confident the changes to the `amd64`, `arm64`, `i386`,
`power` and `riscv` backends are correct. The `arm` backend also
looks good, I am only unsure about the guards mentioning
`v6`, or the lack thereof (maybe a clarifying comment could
be added). My knowledge of `s390x` is far too superficial to be
confident, in particular regarding the number of cycles.

**Comment by xclerc** [comment]
I also think it would be nice to have tests for such changes, even
if it is not easy to run the tests on all architectures. (I remember
using QEMU while working on [1], and it was painful...)

[1] https://github.com/xclerc/ocaml/tree/unsigned_intops_native

**Comment by gretay-js** [comment]
> The `arm` backend also
> looks good, I am only unsure about the guards mentioning
> `v6`, or the lack thereof (maybe a clarifying comment could
> be added). 

I've checked ARMARM for these instructions, and the "unsigned" version is based on the existing signed version. 
ARMv6 introduced only signed mulh (as far as I know) and so I kept umull for all versions of arm.

> My knowledge of `s390x` is far too superficial to be
> confident, in particular regarding the number of cycles.

There is an instruction for unsigned mulh in s390x directly, and it is used by the existing instructions sequence for signed multiply high, so I used the number of issue cycles used for the existing sequence, and reduced in slightly because the new sequence is shorter.

**Comment by gretay-js** [comment]
> I also think it would be nice to have tests for such changes, even
> if it is not easy to run the tests on all architectures. (I remember
> using QEMU while working on [1], and it was painful...)
> 
> [1] https://github.com/xclerc/ocaml/tree/unsigned_intops_native

I have a small CMM test for the new mulh but it requires https://github.com/ocaml-flambda/flambda-backend/pull/291.

**Comment by xclerc** [comment]
> > The `arm` backend also
> > looks good, I am only unsure about the guards mentioning
> > `v6`, or the lack thereof (maybe a clarifying comment could
> > be added).
> 
> I've checked ARMARM for these instructions, and the "unsigned" version is based on the existing signed version.
> ARMv6 introduced only signed mulh (as far as I know) and so I kept umull for all versions of arm.

https://developer.arm.com/documentation/dui0489/h/arm-and-thumb-instructions/umull--umlal--smull--and-smlal
seems to indicate that umull was also introduced in `v6`, doesn't it?


> > My knowledge of `s390x` is far too superficial to be
> > confident, in particular regarding the number of cycles.
> 
> There is an instruction for unsigned mulh in s390x directly, and it is used by the existing instructions sequence for signed multiply high, so I used the number of issue cycles used for the existing sequence, and reduced in slightly because the new sequence is shorter.

That is indeed how I read it, but I don't know if the count for the
existing code was accurate or an approximation, and how your
reduction fares. I am certainly not saying there is anything wrong,
only that I can't vouch for it.

**Comment by gretay-js** [comment]
> https://developer.arm.com/documentation/dui0489/h/arm-and-thumb-instructions/umull--umlal--smull--and-smlal
> seems to indicate that umull was also introduced in `v6`, doesn't it?

According to the [ISA documentation](https://developer.arm.com/documentation/ddi0406/latest ), this instruction has two encodings,  Arm and a Thumb encoding, and the Thumb encoding is the one not available in older versions of the architecture. The availability is the same for both `smull` and `umull`, so the PR does not break anything that hasn't been broken before.

**Comment by xclerc** [comment]
I am always confused with the Thumb/legacy duality, sorry for the noise.

**Comment by gretay-js** [comment]
This PR is ready except a one-line change in `middle_end/flambda2/to_cmm`  that still needs review from one of the code owners please @Gbury @lthls?

**Comment by gretay-js** [comment]
> This PR is ready except a one-line change in `middle_end/flambda2/to_cmm` that still needs review from one of the code owners please @Gbury @lthls?

ping... please @lthls

**Review by lthls — APPROVED** [review]

**Pull request merged**
